File: library-java/lib/src/main/java/eu/postyard/templates/Hashing.java
package eu.postyard.templates;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

/** Hashing helpers used to derive stable, content-addressed cache keys. */
final class Hashing {

  private Hashing() {}

  /**
   * {@return the hex-encoded SHA-256 digest of {@code input}}
   *
   * @param input the string to hash
   */
  static String sha256(String input) {
    try {
      MessageDigest digest = MessageDigest.getInstance("SHA-256");
      byte[] hash = digest.digest(input.getBytes(StandardCharsets.UTF_8));
      StringBuilder hex = new StringBuilder(hash.length * 2);
      for (byte b : hash) {
        hex.append(String.format("%02x", b));
      }
      return hex.toString();
    } catch (NoSuchAlgorithmException e) {
      throw new IllegalStateException("SHA-256 not available", e);
    }
  }
}

File: library-typescript/src/hashing.ts
import { createHash } from "node:crypto";

/**
 * Computes the SHA-256 of a string.
 *
 * @param input - The string to hash.
 * @returns The 64-character lowercase hex digest.
 */
export function sha256(input: string): string {
  return createHash("sha256").update(input, "utf8").digest("hex");
}
